File: modules_metadata/triggers_module.py
#!/usr/bin/python3

#     Copyright 2021. FastyBird s.r.o.
#
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.

"""
Sets of enums for Triggers Module
"""

# Library dependencies
from enum import Enum, unique


@unique
class TriggerControlAction(Enum):
    """
    Control trigger action

    @package        FastyBird:ModulesMetadata!
    @module         triggers_module

    @author         Adam Kadlec <adam.kadlec@fastybird.com>
    """
    TRIGGER_TRIGGER: str = "trigger"

    @classmethod
    def has_value(cls, value: str) -> bool:
        """Check if provided value is valid enum value"""
        return value in cls._value2member_map_  # pylint: disable=no-member


@unique
class TriggerType(Enum):
    """
    Trigger type

    @package        FastyBird:ModulesMetadata!
    @module         triggers_module

    @author         Adam Kadlec <adam.kadlec@fastybird.com>
    """
    MANUAL: str = "manual"
    AUTOMATIC: str = "automatic"

    @classmethod
    def has_value(cls, value: str) -> bool:
        """Check if provided value is valid enum value"""
        return value in cls._value2member_map_  # pylint: disable=no-member


@unique
class TriggerActionType(Enum):
    """
    Trigger action type

    @package        FastyBird:ModulesMetadata!
    @module         triggers_module

    @author         Adam Kadlec <adam.kadlec@fastybird.com>
    """
    DEVICE_PROPERTY: str = "device-property"
    CHANNEL_PROPERTY: str = "channel-property"

    @classmethod
    def has_value(cls, value: str) -> bool:
        """Check if provided value is valid enum value"""
        return value in cls._value2member_map_  # pylint: disable=no-member


@unique
class TriggerConditionType(Enum):
    """
    Trigger condition type

    @package        FastyBird:ModulesMetadata!
    @module         triggers_module

    @author         Adam Kadlec <adam.kadlec@fastybird.com>
    """
    DEVICE_PROPERTY: str = "device-property"
    CHANNEL_PROPERTY: str = "channel-property"
    TIME: str = "time"
    DATE: str = "date"

    @classmethod
    def has_value(cls, value: str) -> bool:
        """Check if provided value is valid enum value"""
        return value in cls._value2member_map_  # pylint: disable=no-member


@unique
class TriggerNotificationType(Enum):
    """
    Trigger notification type

    @package        FastyBird:ModulesMetadata!
    @module         triggers_module

    @author         Adam Kadlec <adam.kadlec@fastybird.com>
    """
    EMAIL: str = "email"
    SMS: str = "sms"

    @classmethod
    def has_value(cls, value: str) -> bool:
        """Check if provided value is valid enum value"""
        return value in cls._value2member_map_  # pylint: disable=no-member


@unique
class TriggerConditionOperator(Enum):
    """
    Condition operator

    @package        FastyBird:ModulesMetadata!
    @module         triggers_module

    @author         Adam Kadlec <adam.kadlec@fastybird.com>
    """
    EQUAL: str = "eq"
    ABOVE: str = "above"
    BELOW: str = "below"

    @classmethod
    def has_value(cls, value: str) -> bool:
        """Check if provided value is valid enum value"""
        return value in cls._value2member_map_  # pylint: disable=no-member

File: public/lib/types/triggers-module.ts
export enum TriggerType {
  MANUAL = 'manual',
  AUTOMATIC = 'automatic',
}

export enum ActionType {
  DEVICE_PROPERTY = 'device-property',
  CHANNEL_PROPERTY = 'channel-property',
}

export enum ConditionType {
  CHANNEL_PROPERTY = 'channel-property',
  DEVICE_PROPERTY = 'device-property',
  TIME = 'time',
  DATE = 'date',
}

export enum NotificationType {
  EMAIL = 'email',
  SMS = 'sms',
}

export enum ConditionOperator {
  EQUAL = 'eq',
  ABOVE = 'above',
  BELOW = 'below',
}

export interface TriggerEntity {
  id: string
  type: TriggerType
  name: string
  comment: string | null
  enabled: boolean
  owner: string | null
  is_triggered?: boolean | null
  is_fulfilled?: boolean | null

  [k: string]: string | TriggerType | boolean | null | undefined
}

export interface TriggerControlEntity {
  id: string
  name: string
  trigger: string
  owner: string | null

  [k: string]: string | null
}

export interface ActionEntity {
  id: string
  type: ActionType
  enabled: boolean
  trigger: string
  device?: string
  channel?: string
  property?: string
  value?: string
  owner: string | null
  is_triggered?: boolean | null

  [k: string]: string | ActionType | boolean | null | undefined
}

export interface ConditionEntity {
  id: string
  type: ConditionType
  enabled: boolean
  trigger: string
  device?: string
  channel?: string
  property?: string
  operand?: string
  operator?: ConditionOperator
  time?: string
  days?: number[]
  date?: string
  owner: string | null
  is_fulfilled?: boolean | null

  [k: string]: string | ConditionType | ConditionOperator | boolean | number[] | null | undefined
}

export interface NotificationEntity {
  id: string
  type: NotificationType
  enabled: boolean
  trigger: string
  email?: string
  phone?: string
  owner: string | null

  [k: string]: string | NotificationType | boolean | null | undefined
}
